bench/workgen/workgen.cpp

```diff
@@ -341,6 +344,48 @@ gen_random_table_name(char *name, workgen_random_state volatile *rand_state)
     }
     name[DYNAMIC_TABLE_LEN - 1] = '\0';
 }
+
+// Create a dynamic table with the specified name and config. If mirroring tables, add the
+// mirror table uri to the table runtime. Return an error status if the uri already exists
+// or the create fails.
+int
+WorkloadRunner::create_table(WT_SESSION *session, const std::string &config, const std::string &uri,
+  const std::string &mirror_uri, const bool is_base)
+{
+    // Check if a table with this name already exists. Return if it does. Use a shared lock to
+    // read the dynamic table structure.
+    ContextInternal *icontext = _workload->_context->_internal;
+    {
+        const std::shared_lock lock(*icontext->_dyn_mutex);
+        if (icontext->_tint.count(uri) > 0 || icontext->_dyn_tint.count(uri) > 0)
+            return EEXIST;
+    }
+
+    // Create the table.
+    WT_DECL_RET;
+    if ((ret = session->create(session, uri.c_str(), config.c_str())) != 0) {
+        if (ret != EBUSY)
+            THROW("Failed to create table '" << uri << "'.");
+        return ret;
+    }
```

Comment:
I'm going to call on @sulabhM to get his thoughts as he did the original implementation for this code and spent a fair amount of time on the locking in particular.

Reply:
@ddanderson , @tammybailey - yes I am keeping the slow operation outside the lock. This is a single thread that does creates, so we don't have to worry about anyone else getting in the way. The first lock to check the tables is the shared lock - since it can co-exist with the operation threads that take the lock on the structure when selecting a table to operate on. The next lock is the exclusive lock that blocks the other shared locks in the meanwhile.

In short, this thread won't race with anyone else creating tables - so we are good. The two part shared lock - exclusive lock is to minimize readers having to wait longer than needed and keeping slow schema operations outside the lock.

Reply:
Ah, thanks @sulabhM.  That makes sense.